lib/route/act/nat.c

```diff
@@ -0,0 +1,303 @@
+/*
+ * lib/route/act/nat.c	NAT action
+ *
+ *	This library is free software; you can redistribute it and/or
+ *	modify it under the terms of the GNU Lesser General Public
+ *	License as published by the Free Software Foundation version 2.1
+ *	of the License.
+ *
+ * Copyright (c) 2016 Magnus Öberg <magnus.oberg@westermo.se>
+ */
+
+/**
+ * @ingroup act
+ * @defgroup act_nat NAT
+ *
+ * @{
+ */
+
+#include <netlink-private/netlink.h>
+#include <netlink-private/tc.h>
+#include <netlink/netlink.h>
+#include <netlink/attr.h>
+#include <netlink/utils.h>
+#include <netlink-private/route/tc-api.h>
+#include <netlink/route/act/nat.h>
+#include <netlink/route/tc.h>
+
+static struct nla_policy nat_policy[TCA_NAT_MAX + 1] = {
+	[TCA_NAT_PARMS] = { .minlen = sizeof(struct tc_nat) },
+};
+
+/**
+ * nat operations
+ */
+
+static int nat_msg_parser(struct rtnl_tc *tc, void *data)
+{
+	struct tc_nat *nat = data;
+	struct nlattr *tb[TCA_NAT_MAX + 1];
+	int err;
+
+	err = tca_parse(tb, TCA_NAT_MAX, tc, nat_policy);
+	if (err < 0)
+		return err;
+
+	if (!tb[TCA_NAT_PARMS])
+		return -NLE_MISSING_ATTR;
+
+	nla_memcpy(nat, tb[TCA_NAT_PARMS], sizeof(*nat));
+
+	return NLE_SUCCESS;
+}
+
+static void nat_free_data(struct rtnl_tc *tc, void *data)
+{
+}
+
+static int nat_clone(void *_dst, void *_src)
+{
+	struct tc_nat *dst = _dst, *src = _src;
+
+	memcpy(dst, src, sizeof(*src));
```

Comment:
this clone function only does a plain memcpy. I don't think you need to implement it, the default implementation should do this already. Please check whether this function can be dropped.